### ISSUE TITLE
Accessing a private view/project through URL without credentials

### ISSUE THREAD

**Issue by antobenve** [unep-grid/map-x-mgl#635]
- [x] Accessing a private project using the "views" parameter -> no views if user opts to remain in Home
If the user opts to remain in Home, views of Home should be loaded.
Please mind that the issue is related to the use of the "views" parameter ([example](https://app.staging.mapx.org?project=MX-00H-Q47-ZWW-13W-6ON&views=MX-JJBGP-70I42-0SG3L&viewsOpen=MX-JJBGP-70I42-0SG3L&language=en&
)). [Without it](https://app.staging.mapx.org?project=MX-00H-Q47-ZWW-13W-6ON), no issue.

- [x] Accessing a private view for which the user doesn't have the rights -> no views are loaded
If the user has the rights to access the project but not the view, no view is loaded in panel ([example](https://app.staging.mapx.org?project=MX-BDS-AK2-PKN-ZJQ-KXK&views=MX-9A8TQ-FIL23-IS6TT&viewsOpen=MX-9A8TQ-FIL23-IS6TT&language=en&)). A potential solution is a message saying "The view/s that you're trying to open are private and cannot be loaded with your account" together with a button to load the content of the project where the user is.

**Private or public view in private project:** 
- [x] The message "The view is unavailable" is received before the user gets the chance to login. -> MapX should not inform the user that the view does not exists in the project until the user has tried to login. 

**Private view in public project:**
- [ ] The message "The view is unavailable" is received + the user is not given the change to login in order to reach the view. -> MapX should propose the user to login. + MapX should not inform the user that the view does not exists in the project until the user has tried to login. 

Edited on 05/07/2021 to list the bugs that should be fixed in Summer 2021.

**Comment by antobenve** [comment]
I just edited the post above as I realized my description of the two issues was very poor. Is it clear now?

**Comment by fxi** [comment]
Until now, when a private project with selected views was requested, but the user was not logged in, the default project was displayed and the selected views were requested in that project. If the view was present, in this project - as an external view or as a local view - an empty list view was returned.

You requested the default list of views for this project to be loaded. So ... I patched the API to make a second attempt: if a filter is defined and no views are returned, the default list is now returned. MapX has no way of knowing the views of a project in advance, as it depends on many factors.

For the second point, I added a modal that explains that views were requested, but not available: we can't know for sure that the view was not deleted 3 months ago, or user n not allowed to read it. So, it's just "unavailable".

![image](https://user-images.githubusercontent.com/1196833/105359108-5a99a300-5bf7-11eb-89c9-1ccad4a97ca5.png)

I've removed a list of views id and titles, because ids are not nice to display and titles could reveal unwanted information. In the case of a private views, it's not what we want.  But if the user has a view ID, it can load the view in static mode....

**Comment by antobenve** [comment]
For the first point, I think that the chances that one of the views that were requested in the project X is available in the project Home is so remote that is way better to uniformize the behavior of MapX when a user access a private project (for which the user doesn't have credentials) with or without views selection. 
For the second point the modal is fine with saying "the view is unavailable". At this point the user should be allowed to load something - which is the content of the project that was loaded. Is "close" allowing this?

**Comment by fxi** [comment]
Please test the new behaviour in a live app (1.8.5-beta)  and report next.

**Comment by fxi** [comment]
```json
{
   "id": "fetch_views_diff_message",
    "en": "You have requested views that are not available in this project.",
    "fr": "Vous avez demandé des vues qui ne sont pas disponibles dans ce projet.",
  }
```

could be 

```json
{
   "id": "fetch_views_diff_message",
    "en": "You have requested views that are not available in this project. All available views for your are displayed.",
    "fr": "Vous avez demandé des vues qui ne sont pas disponibles dans ce projet. Toutes les vues disponibles pour vous sont affichées.",
  }
```
That's a generic message. It will be displayed for all cases when a view is missing - using the fetchView method  at least –, not only: private project + private view.

If we handle each edge cases with a custom logic, the app will probably gain too much complexity.

**Comment by antobenve** [comment]
You're right. Best is the way it is written as long as clicking "close" the user loads the content of the project for which he has the rights to.

However, it would be best if the message does not overlay the login window that opens when the user doesn't have the rights to access a project. 
To me, if easy of implementation> 
- user can't access project >  Only show that the user must login with a valid account to open the project.
- user can access project >  Check if any view is selected in URL > if so, provide the new message to the user and load the content of the project when the user clicks "Close".

**Comment by antobenve** [comment]
@fxi While the first point is 100% done, the second one could be improved following my before-last comment. Should we leave the issue open and downgrade it to priority 2? Should I reformulate the text in the issue too?

**Comment by antobenve** [comment]
Hello, I just realized that the app misbehaves when I try to access a private view in a private project and I'm not logged in yet.

- The message "The view is unavailable" is received before I get the chance to login. -> MapX should wait to check if the view exists in the project because the right project is not loaded yet

- Furthermore, right after I log in, the message is sent again and the panel is loaded with views from the "Home" project instead of those of the project that I managed to open thanks to my log in.
